receiver/k8sclusterreceiver/watcher.go

```diff
@@ -35,23 +37,28 @@ import (
 )
 
 type resourceWatcher struct {
-	client                kubernetes.Interface
-	sharedInformerFactory informers.SharedInformerFactory
-	dataCollector         *collection.DataCollector
-	logger                *zap.Logger
-	metadataConsumers     []metadataConsumer
+	client                  kubernetes.Interface
+	sharedInformerFactory   informers.SharedInformerFactory
+	dataCollector           *collection.DataCollector
+	logger                  *zap.Logger
+	metadataConsumers       []metadataConsumer
+	initialCacheSyncTimeout time.Duration
+	informersHaveSynced     *atomic.Bool
```

Comment:
Why it has to be a boolean flag? Can't we synchronize it with a channel which would block `kr.resourceWatcher.startWatchingResources` until the cache is synced?

Reply:
I tried that approach initially but I learnt that `informers.SharedInformerFactory.Start` needs to be invoked first for the cache sync to begin. And as soon as `Start` is invoked events could come through to the Resource Event Handler before the caches are synced. So in this particular case the receiver would have to invoke `rw.sharedInformerFactory.Start(stopper)` to start kick-off the sync and ensure any events that come through before the initial sync is complete, blocks until the initial cache sync is complete or it times out.